### ISSUE TITLE
Comprehensive 6502 testsuite run on sim65 exposes a number of implementation bugs.

### ISSUE THREAD

**Issue by sidneycadot** [cc65/cc65#2539]
Hello,

After an initial report on a simulator bug (#2536), and following a suggestion from user @kugelfuhr in that issue's discussion, I implemented a program to test the "sim65" simulator against the comprehensive "65x02" suite of 6502 instruction behavior tests (https://github.com/SingleStepTests/65x02).

The repository containing the sim65-test program can be found here:

https://github.com/sidneycadot/sim65-test-prototype

This test exposes several bugs in sim65, among which the original bug as reported in #2536, but also several overflow bugs in the simulator code, as well as several issues with the emulation of other 6502 instructions. For example, the tests indicate that the cycle counts of branch instructions that do page-crossing are not correct in some circumstances.

I'd like to discuss several issues here:

* Is someone in particular the go-to person for issues with the sim65 simulator at this point in time?

* What is the best way to handle the issues exposed by these tests? Should we make a single issue-tracker issue for each type of behavioral discrepancy, or should someone make a dedicated effort to minimize or eliminate the discrepancies between sim65 behavior and the testsuite behavior?

* To what extent do we trust the testsuite? At first sight it looks well done, but there is no real traceability on how the test cases were created. We can either decide to trust the testsuite as is, and strive for compliance; or, in every instance where the behavior of sim65 deviates from the behavior described in the testsuite, investigate what happens on a real (physical) 6502. The latter is potentially quite time-consuming.

* It would be possible to integrate the sim65-test program (or something like it) into the cc65 project; perhaps as a stand-alone test program, or as a part of the standard test procedure, or as part of the sim65 program (an idea by user @kugelfuhr, not necessarily my preference). Would that be desirable? Keeping sim65-test as a separate project is also certainly an option, I think.

* The sim65 simulator code (6502.c, in particular the 'ExecuteInsn' function) is pretty nice; however, it is also obvious that it is not written to be used in a context other than running the sim65 program. I had to deal with / cut several implicit dependencies of the 6502.c code to the sim65 code around it (see the "How it works" section in https://github.com/sidneycadot/sim65-test-prototype/blob/main/README.md). It would perhaps be possible to make slight changes to the sim65 code to make it easier to test its core 'ExecuteInstn' routine.

**Comment by sidneycadot** [comment]
To make all this a bit more concrete, here's an overview of the test results of sim65 vs the 65x02 testsuite.

![image](https://github.com/user-attachments/assets/d38bb965-7227-429d-8936-79505ab6e420)

**Comment by sidneycadot** [comment]
![image](https://github.com/user-attachments/assets/217caaf8-0aae-4762-92d0-9643a71724c9)

Right, so I digged in and I now have a version that is fully in line agreement with all cases in the 65x02 test suite. It was indeed around 10 bugs that needed fixing.

The most obscure bug (if you can call it that) was a test case where a JSR instruction was located inside the page 1 stack, and the target address of the JSR is altered _as the return address is being pushed_. That's a case that's easy to miss without a massive test set :)

I think it will be worthwhile to get at least the bugfixes into sim65, but we still need to discuss the points mentioned in my original report, and I have some additional questions regarding coding style.

Most importantly, I want to know if it's okay to include <inttypes.h> and use its uint8_t and uint16_t types, which help to fix a slew of nasty issues that occur because the 8- and 16-bit registers of the CPU are currently all represented as unsigneds (which are 32 bits on most modern compilers). It is very easy to miss cases where assignments to a register are out-of-range because of that. Assigning to uint8_t or uint16_t takes care of the proper bit-masking essentially for free, so it's really a big improvement.

**Comment by sidneycadot** [comment]
Note: I closed #2536 as resolving this issue will also take care of the issue reported there.

**Comment by sidneycadot** [comment]
The lack of response to my questions is worrying (and, frankly, annoying).

I am quite willing to put in the work to prepare a pull request that fixes all sim65 problems detected, but if it just ends up in the pull request queue with the other 23 pending pull requests, that would just be a waste of my time.

I need to see some commitment that this will be taken seriously.

**Comment by kugelfuhr** [comment]
Maybe you should tell whom you are addressing. I am not a cc65 developer and anything I had to say was said in #2536.

**Comment by mrdudz** [comment]
Sorry, my time budget for cc65 is very limited these days - i try to check pending issues at least once per week, but sometimes might not be able to do even that. So please be patient - i AM reading everything that is posted, and i WILL react. However it may take a bit :)

> Is someone in particular the go-to person for issues with the sim65 simulator at this point in time?

No, not really. I am usually the one who would merge the PR - but there are certainly others around who know this code better than i do - you might want to check the commit history, for example. For example, someone recently forked sim65 to hack it and run (part of) the "Lorenz testsuite" (included in the VICE testbench) - which also found some bugs.

> What is the best way to handle the issues exposed by these tests? Should we make a single issue-tracker issue for each type of behavioral discrepancy, or should someone make a dedicated effort to minimize or eliminate the discrepancies between sim65 behavior and the testsuite behavior?

Since you already fixed the problems... This ticket is fine, i guess :) (But usually - one ticket per issue please, unless they are so closely related that splitting it does not make sense)

> To what extent do we trust the testsuite? At first sight it looks well done, but there is no real traceability on how the test cases were created. We can either decide to trust the testsuite as is, and strive for compliance; or, in every instance where the behavior of sim65 deviates from the behavior described in the testsuite, investigate what happens on a real (physical) 6502. The latter is potentially quite time-consuming.

That is a bit of a problem, indeed - a testbench that isn't verified on the real CPU is questionable at best. So i would indeed prefer to verify any such fixes on the real thing. That's one reason for why on my TODO list in the Wiki it mentions support for (part of) the VICE test suite - those programs are all verified on a real C64 (by me actually). Making this happen is still a bit of work however (and might sometimes not even be worth it). Again my time budget is the problem here, if i had the time for it, i'd already had done this. I might be able to point to the specific tests though, if you can list the ones from the above testbench that failed (And i like to believe i know the 6502 well enough that i can "verify" some of them by looking at them).

> It would be possible to integrate the sim65-test program (or something like it) into the cc65 project; perhaps as a stand-alone test program, or as a part of the standard test procedure, or as part of the sim65 program (an idea by user @kugelfuhr, not necessarily my preference). Would that be desirable? Keeping sim65-test as a separate project is also certainly an option, I think.

I don't like the idea to have this stuff in the regular sim65 - it's too specific. I'd much prefer a similar approach as used by the VICE testbench, IF we are going this route. A separate project (that pulls in sim65 as a sub project or sth like that) sounds like a good idea to me.

> The sim65 simulator code (6502.c, in particular the 'ExecuteInsn' function) is pretty nice; however, it is also obvious that it is not written to be used in a context other than running the sim65 program. I had to deal with / cut several implicit dependencies of the 6502.c code to the sim65 code around it (see the "How it works" section in https://github.com/sidneycadot/sim65-test-prototype/blob/main/README.md). It would perhaps be possible to make slight changes to the sim65 code to make it easier to test its core 'ExecuteInstn' routine.

Changes that make it easier are fine with me IF they don't affect the normal behaviour and/or do not make the implementation more complicated/complex - any such thing should go into that separate repo then.

(*) https://vice-emu.pokefinder.org/wiki/Testbench - in particular https://sourceforge.net/p/vice-emu/code/HEAD/tree/testprogs/CPU/ and https://sourceforge.net/p/vice-emu/code/HEAD/tree/testprogs/general/Lorenz-2.15/

**Comment by sidneycadot** [comment]
Hi @kugelfuhr, this is addressed to whomever is doing issue-tracker triage; I guess that's @mrdudz these days. It was not directed at you; your comments and suggestions from #2336 are much appreciated.

**Comment by sidneycadot** [comment]
Hi @mrdudz thanks for the answers. I understand that you're busy, but not a single peep after five days was a bit much.

I am unfamiliar with the VICE testset. About half the bugs I fixed were clockcycle-counts being off by 1 (if the testsuite is to be believed); I don't know if the VICE tests include 6502 clock cycle tests (?)

It would be very helpful if there were a somewhat comprehensive testsuite that can be run on actual hardware, for cross-comparison with the simulator. For the basic opcode functionality that's quite doable, I think; however, establishing the cycle-accurate timing characteristics of all variations of instructions is quite a lot harder to measure on physical machines, first because 6502 hardware is not usually equipped with clockcycle counter circuitry; and second because (even if you disable interrupts and DMA) the CPUs of the old 8-bit machines are sometimes halted to insert memory refresh cycles.

If such an effort was already tried for the C64, we could indeed piggy-back on that effort. If not, I will have to think real hard if I can somehow make this work on an Atari. I think I may see a way (with a nice trick using the Atari's hardware-based sprite collission detection) but it will surely take significant effort to implement.

Such a test would provide the ultimate reference for sim65 (and other emulators), at least with respect to behavior of the 6502.

If we decide now that we don't fully trust the 65x02 testsuite (which is very conservative, but entirely reasonable), I should probably suspend work on fixing sim65 at this point and re-focus efforts to get such a comprehensive test-suite working, including an instruction timing characterization on Atari hardware. Unless, of course , there is an existing project that does precisely that on the C64 or some other machine, or if there is a gold standard instruction count test that we do trust.

One problem is that this would only cover the 6502, not the other 6502 variants supported, the 65c02 and the 6502x (whatever that is).

> A separate project (that pulls in sim65 as a sub project or sth like that) sounds like a good idea to me.

I agree. I personally also do not like to encumber sim65 with test-specific functionality; one of the appeals of the simulator as it is, is that it is small, and its source code (once you know how to read the many macros) is pretty straightforward.

> Changes that make it easier are fine with me IF they don't affect the normal behaviour and/or do not make the
> implementation more complicated/complex - any such thing should go into that separate repo then.

There are three things I'd like to change:

(1) Make the "Regs" variable that keeps the CPU state non-static, so an external test routine can access it without shenanigans.
(2) Provide a way to disable the ParaVirtual hooks (used by the sim65 to access stuff like files in the external program).
(3) Provide a way to disable the process exit when encountering an illegal opcode, and the error/warning messages when encountering illegal instructions and the JMP (indirect) bug.

None of this should change the behavior of the program when run normally.

Ultimately (if the comprehensive test suite including instruction cycle counts were available), such a test could run completely self contained inside sim65, at least if the feature request #2355 were implemented.

**Comment by mrdudz** [comment]
> I don't know if the VICE tests include 6502 clock cycle tests (?)

Unfortunately only for some selected (illegal) opcodes ... the Lorenz Suite only checks if the opcodes do the right thing, not their timing (which would be tricky the way it works)

> however, establishing the cycle-accurate timing characteristics of all variations of instructions is quite a lot harder to measure on physical machines, first because 6502 hardware is not usually equipped with clockcycle counter circuitry; and second because (even if you disable interrupts and DMA) the CPUs of the old 8-bit machines are sometimes halted to insert memory refresh cycles.

On the C64 this is no problem - disable the screen and use a CIA timer to measure the number of cycles. I actually happen to have written a "Benchmark" that does exactly this (it is even made with cc65 :)) - see https://github.com/mrdudz/synthmark/tree/master ( https://github.com/mrdudz/synthmark/blob/master/main.c#L93 ). If you run the instruction you want to test 10 or 100 times, its very easy to spot deviations from the expected value, even if the framework code isn't 100% cycle exact (but this is not hard to do either)

> If we decide now that we don't fully trust the 65x02 testsuite (which is very conservative, but entirely reasonable), I should probably suspend work on fixing sim65 at this point and re-focus efforts to get such a comprehensive test-suite working

I wouldn't go that far, as said before - in particular regarding the timing i'd stick to cross-checking the test that fail. If only because the cyclecount being wrong isn't really a major problem - much less than correctly working opcodes anyway.

> 6502x (whatever that is).

That is regular 6502, plus illegal opcodes (which the regular 6502 mode in the assembler does not support)

> There are three things I'd like to change: [...]

sounds entirely reasonable to me

**Comment by sidneycadot** [comment]
>> If only because the cyclecount being wrong isn't really a major problem - much less than correctly working opcodes anyway.

I see your point, but still, if clock cycles are counted and reported, they should be correct; or alternatively, the tool's documentation should explicitly say that they are merely indicative.

I suggest the following course of action:

- I will first prepare a first pull request that introduces the sized types (uint8_t, uint16_t) which already solve some problems.
- I will prepare a second pull request covering the three (configurable) behavioral changes I proposed above that will aid in further testing of the ExecuteInstn code.
- I will then preparing a third and last PR that brings the sim65 behavior in line with the 65x02 testsuite. I will add a comment to the code that the opcode actions and cycle counts have been tested against the 65x02 test suite, and that they are as correct as that test suite is. That provides at least some degree of tracability, even if it is to a testsuite that itself does not provide much in terms of traceability to its result. That's an improvement relative to the current situation where there is no traceability of correctness at all. I will make an explicit list of changes (both in terms of opcode functionality and cycle counts) of all the changes I make, so they can be tested by anyone who feels the urge. I am not a C64 guy so it would be inconvenient for me to run those tests. It is up to you (@mrdudz) to decide if you're willing to accept the PR without that work, or if you will want to see tests for those cases first (or perhaps a subset, to gain confidence).

**Comment by rofl0r** [comment]
a good alternative to using real hardware for cycle counts should be perfect6502/visual6502 btw. since those are emulating the hw on a transistor level. i've got a [fork](https://github.com/rofl0r/visualz80) of visual6502 which is by far easier to build than upstream, though nothing speaks against using the [wasm build](https://github.com/floooh/v6502r) in a browser.

**Comment by mrdudz** [comment]
3 incremental PRs sound like a great idea to me.

And yes, i will verify the fixes. As said already, i might be able to do this by looking at the testcases - since i am quite familiar with 6502 cycle counts and quirks :)

**Comment by sidneycadot** [comment]
Ok, I will prepare the first PR somewhere over the weekend.

Meanwhile, I figured out a pretty spiffy way to do instruction cycle counting on the Atari without interference from the memory refresh cycles. Using CC65 I can now write this:

```
///////////////////////////
// test_measure_cycles.c //
///////////////////////////

#include <stdio.h>
#include <stdint.h>

#include "measure_cycles.h"

uint8_t code[4096];

int main(void)
{
    uint16_t cycles;

    code[0] = 0xea; // nop
    code[1] = 0x60; // rts (will not be included in timing).
 
    dma_and_interrupts_off();

    cycles = measure_cycles(code);

    dma_and_interrupts_on();

    printf("result: %u cycles.\n", cycles);

    while (1);

    return 0;
}
```

Lo and behold, this actually works -- in emulation, but should also work on real hardware:

![image](https://github.com/user-attachments/assets/210a907e-1cc9-4f13-afaf-87663cc45f10)

This is hopefully the starting point for a pretty thorough clock-cycle hardware based clockcycle tester that can now be written in C thanks to cc65 :-)

**Comment by mrdudz** [comment]
Please make sure it has the atari specific part separate so it can be ported to c64 - would be a great addition to the VICE tests :)

**Comment by sidneycadot** [comment]
As to those tests: will do. I already found a friend who is willing to do a C64 version. Will be fun.

As to the PR I'm preparing: I hit a snag.

This first pull request intends to change the type of the PC registers and memory to 'uint8_t' and 'uint16_t', which gets rid of a handful of overflow bugs.

The problem is that several of the opcode implementations in 6502.c depend subtly on 8-bit registers (temporarily) holding values beyond their lower 8 bits. Obvious examples are the implementations of ADC, SBC, ROL, and ASL.

Now I can fix those 6502 opcode implementations without too much issue.

The problem is that there are also 65C02 and illegal opcodes that do the same. For those, I don't have the knowledge to fix them (and, frankly, I also don't have the motivation -- I don't really care all that much to fix those opcodes). Another problem is that we lack an automated testset, to see if those re-implementations break stuff; this is especially a problem for the 65c02 and the 6502x variants (for the 6502 I'm comfortable I could do the tests using my sim65 test).

So this is a bit of a rock and a hard place. To summarize:

- We know that there are currently bugs in 6502.c.
- An important step in fixing those is switching to properly sized types for the CPU registers.
- This requires work on the opcode implementations, which can potentially break stuff.
- We don't have a proper way to test that we don't break stuff, especially for the 65c02 and 6502x variants.

Perhaps the conclusion here is once again that we first need proper test infrastructure for the simulator, that covers all supported variants, before we start changing things that can break stuff.

**Comment by kugelfuhr** [comment]
Why not use the 65c02 and 6502x tests from the SingleStepTests? I don't know good the tests are for the 6502 illegals, but using the 65c02 tests should be pretty safe. I didn't have a thorough look at your implementation, but just choosing a new input file and another CPU should be too much work?

**Comment by kugelfuhr** [comment]
You do not have to fix the 6502x and 65c02 stuff. Just make sure it behaves the same as before. So
- run the 65c02 tests and keep the results,
- make your changes regarding data types to the 65c02 instructions,
- run the tests again,
- and finally make sure the results are the same.

This way you can make your changes without having to fiddle with stuff you don't want to. Yes, you will have to touch the 65c02 instruction code, but leaving it alone and changing data types only for 6502 instructions would create a mess. But I think it's ok not to fix these if they don't behave correctly.

You could even take this somewhat further changing the order of the steps:

- First add your tests. This documents the current state and shows which instructions have errors.
- Next, fix the data types making sure that the test results do not change.
- Finally fix the 6502 instructions and leave 6502x and 65c02 to others.

**Comment by sidneycadot** [comment]
> Just make sure it behaves the same as before.

That's pretty much impossible without significant effort if I change the register types from "unsigned" to "uint8_t".

I noticed that you have a habit of saying stuff like "that's not much work", or "just do X". It often is not as easy as you make it out to be, and it is a bit grating to be on the receiving end of that. I'd appreciate if you didn't do that.

**Comment by mrdudz** [comment]
I would NOT trust the tests regarding illegal opcodes - at all. There have been a lot of myths and half truths about those "in the wild", and a lot of existing implementations get some of the details wrong. You also can't verify all of them on visual6502 (or perfect6502 for that matter), because it doesn't simulate the analog side effects that contribute to their behaviour. (I spent a lot of time researching those -> see https://csdb.dk/release/?id=238036 if you are curious)

> The other way forward could be that we accept that 65C02 and 6502X support will be broken by these fixes. This would prioritize correctness of the 6502 over support for the other variants. But I don't think that's a wise course of action.

That sounds like a really bad idea, we should never knowingly break something :)

I'd probably try to explicitly AND with 0xFF at the proper places to fix the broken opcodes at this point

**Comment by sidneycadot** [comment]
> That sounds like a really bad idea, we should never knowingly break something :)

I agree that it's a bad idea, but at this point I'm struggling to see a good way forward. 

> I'd probably try to explicitly AND with 0xFF at the proper places to fix the broken opcodes at this point

So do you suggest leaving the register types as 'unsigned', and not changing them to 'uint8_t', 'uint16_t' ?

I think that's a lot of work, it's error-prone (especially without a heavy-duty test set in place), and it is a kicking-the-can-down-the-road approach -- the unfortunate choice of typing all registers as bare 'unsigned' should be fixed at some point in time, and this is as good a time as any.

**Comment by kugelfuhr** [comment]
> I noticed that you have a habit of saying stuff like "that's not much work", or "just do X". It often is not as easy as you make it out to be, and it is a bit grating to be on the receiving end of that. 

That's difficult in cases where I know how much work or how difficult it is since I have done almost exactly these things. As an example look at [your reply to my suggestion to use the tests](https://github.com/cc65/cc65/issues/2536#issuecomment-2478731390) and my answer.

Would you have started writing the test without me saying it's not that much work?

Anyway, I will try to be more careful in the future.

**Comment by sidneycadot** [comment]
You may know how much work it is for you; but you don't know how much work it is for me. Perhaps you're much faster than I am, or perhaps you underestimate the amount of work involved because you didn't see some nitty-gritty details that I did. And there's also the matter that "much" is entirely subjective.

Anyway, thanks for your consideration.

**Comment by mrdudz** [comment]
> So do you suggest leaving the register types as 'unsigned', and not changing them to 'uint8_t', 'uint16_t' ?
>
> I think that's a lot of work, it's error-prone (especially without a heavy-duty test set in place), and it is a kicking-the-can-down-the-road approach -- the unfortunate choice of typing all registers as bare 'unsigned' should be fixed at some point in time, and this is as good a time as any.
> 
> I'm still hoping we can think of a better way.

It really depends on how many such changes would be needed - if it's just a few, then i don't see why not. If it's all over the place, then yes, please don't :)

**Comment by sidneycadot** [comment]
It's quite a few places including in opcode handlers that I don't want to touch (like the 6502X and 65C02) in the absence of proper regression testing infrastructure that actually tests those opcodes.

This unfortunately means that we've reached a deadlock on how to make progress on this issue.

The proper way to break the deadlock would be to select some test suite as an authoritative reference, making sure that sim65 can be run against it easily, and then start fixing stuff, including stuff in the 65C02 and 6502X CPU variants. Ideally, such a test set would include both functional tests and timing information. 65x02 would fit the bill very well, except for the fact that it doesn't provide proper traceability to real hardware, or any traceability at all, in fact.

Unless a new idea to make progress comes soon I will just disengage from this issue. It's a bit frustrating -- on the one hand we can point out real, actual bugs in sim65; on the other hand, there is no obvious path to fixing them, due to the lack of regression testing. I think that the latter is the root issue here, and it will need to be solved first.

**Comment by mrdudz** [comment]
> The proper way to break the deadlock would be to select some test suite as an authoritative reference, making sure that sim65 can be run against it easily, and then start fixing stuff, including stuff in the 65C02 and 6502X CPU variants. Ideally, such a test set would include both functional tests and timing information. 65x02 would fit the bill very well, except for the fact that it doesn't provide proper traceability to real hardware, or any traceability at all, in fact.

Perhaps you could just make a PR with your fixed version... afterall, the CI will run a ton of little test programs, which is unlikely to still work if something is seriously broken. Its not an ideal situation, i agree. If anything, it would help to see what and how much would be changed/touched though.

**Comment by sidneycadot** [comment]
Perhaps. Can I easily reproduce the CI tests locally?

**Comment by mrdudz** [comment]
just do "make test" in the toplevel of the tree

**Comment by mrdudz** [comment]
make needs to (and does, with git head) complete without error, then everything works as expected (that includes a bunch of tests failing). The last error you probably see is some sprintf test that failed, but that is expected. but as said, make should NOT error out (you might want to try what happens with an unmodified tree)

That said, the CI can also fail when a test that is expected to currently fail (usually in the TODO subdirectory) is now working (for whatever reason). In any such case a PR should also move that test from TODO to "err" (most likely, see the readme in the test subdir). I'd be very surprised if that is the case here though - it would mean that something that we thought is a compiler problem is actually caused by broken sim65.

**Comment by sidneycadot** [comment]
Ok I re-checked with a pristine clone and this time the testsuite completes without issue. Strange, as the clone I ran the tests in earlier had a squeaky clean 'git status', so I'm not sure what's different.

I will start spending time on making a PR, but no promises as to when this will come -- could be days, could be weeks or even months.

**Comment by mrdudz** [comment]
No hurry.... its not like a lot of other development is going on right now :)

**Comment by sidneycadot** [comment]
I'm examining the situation of the 6502X and 65C02 CPU variants. Comparing to the version I made that fixes all 6502 bugs (when taking the 65x02 testsuite as a ground truth) the situation looks not too shabby. 


Ignoring clock-cycle exactness for the time being (let's address those once we have the basic behavior right), there's 7 6502X opcodes that show differences between my "fixed-sim65" behavior and the 65x02 testsuite; and there's 19 65C02 instructions that show issues. The 65C02 issues seem clustered, there's probably just a handful of underlying issues to be solved.

I can visit those issues one by one and (even without a lot of experience with 6502 illegals and the WC6502) I should be able to pinpoint the behavioral differences.

One question is open that would help to focus my work. The 65x02 testsuite supports three sub-variants of the 65C02, by manufacturer: there's a Rockwell version, a WDC version, and a Synertek version; the three of them show super subtle differences in the processing of some opcodes, and some small differences in instruction timing.

I think it would be good to pick a 65C02 variant that sim65 aims to support. I feel the WDC 65C02 is probably the most obvious candidate to use as a reference. Given that I will get access to a WDC6502 soonish that I can use for testing, I would prefer if we declare the WDC 65C02 as the reference chip.

Any opinions?

**Comment by sidneycadot** [comment]
As to timing testing (to be fixed after getting the behavior correct): I made a pretty nice tool that verifies instruction timings on 6502 instructions. It currently supports the 151 official opcodes, and it runs on hardware 6502 (Atari 8-bit and C64 at the moment), so it provides the desirable traceability to hardware.

Have a peek at: https://github.com/sidneycadot/6502-test

I want to include sim65 as a target for that tool as well, to enable traceability of its timing behavior to hardware. To that end, I will revive the discussion at issue #2355 which has stalled at the moment.

**Comment by mrdudz** [comment]
> I think it would be good to pick a 65C02 variant that sim65 aims to support. I feel the WDC 65C02 is probably the most obvious candidate to use as a reference. Given that I will get access to a WDC6502 soonish that I can use for testing, I would prefer if we declare the WDC 65C02 as the reference chip.

I have no opinion about 65C02, zero experience with that. I would take a variant that was used in one of the targets we support (thats most likely not the WDC one)

As for regular 6502 - i would NOT use the WDC one as a reference, but stick to the original MOS one (or Synertek or Rockwell - which used the exact same mask and work exactly the same)

>  there's 7 6502X opcodes that show differences between my "fixed-sim65" behavior and the 65x02 testsuite;

Which opcodes are that? 7 is a suspicious number - its exactly how many so called "unstable" illegal opcodes exist (which many existing emulations get wrong - stick to the above mentioned document for those :))

**Comment by sidneycadot** [comment]
>  I would take a variant that was used in one of the targets we support 

Can you be more specific? I am just about as experienced as you are with the 65c02 , namely, not (yet). I do have a neo 6502 come in which uses a WDC 6502, so there is that.

> As for regular 6502 - i would NOT use the WDC one as a reference

Sorry that was a typo.

There's no ambiguity for the regular 6502, the 65x02 testsuite has only 1 variant for that.

> Which opcodes are that? 7 is a suspicious number - its exactly how many so called "unstable" illegal opcodes exist (which many existing emulations get wrong - stick to the above mentioned document for those :))

![image](https://github.com/user-attachments/assets/5854ff2c-e225-4829-b39f-70717179bcd6)

> stick to the above mentioned document for those :))

I want to avoid making a study of the illegal opcodes; I want to prioritize getting the legal opcodes to work correctly.

**Comment by mrdudz** [comment]
> Can you be more specific? I am just about as experienced as you are with the 65C02 , namely, not (yet).

The toolchain supports these (from target.c)

    { "apple2enh",      CPU_65C02,      BINFMT_BINARY,      CTNone  },
    { "geos-apple",     CPU_65C02,      BINFMT_BINARY,      CTNone  },
    { "cx16",           CPU_65C02,      BINFMT_BINARY,      CTPET   },
    { "rp6502",         CPU_65C02,      BINFMT_BINARY,      CTNone  },

and:
   { "supervision",    CPU_65SC02,     BINFMT_BINARY,      CTNone  },
    { "lynx",           CPU_65SC02,     BINFMT_BINARY,      CTNone  },

I have no experience with any of these (or know what exact CPU they use), nor any idea what the difference between 65C02 and 652SC02 is. My guess would be that apple2 used either rockwell or synertek (ie both), and cx16 and rp6502 use WDC.

That probably means its fine to pick whatever you think is fine :) The differences should be negliable anyway - just document what its supposed to be. If you can check against a real CPU, that seems a clear vote for the one you can check against :)

> I want to avoid making a study of the illegal opcodes; I want to prioritize getting the legal opcodes to work correctly.

Oh, sure :) In that table,all orange ones are unstable illegal ones. Interestingly $ab passes all tests :) I will chime in and fix those later, perhaps.

(And that means all the "stable" ones seem to work OK - good)

**Comment by sidneycadot** [comment]
Ok, then I'll proceed with the WDC 65C02 as baseline. I agree with your assessment that differences between the 65C02 variants will probably be near-negligible; and once someone with an interest, ability, and access to the different 65C02s comes along it will be easier for them to work out the differences against a known-good implementation anyway (fat chance that will ever happen, of course).

I'll start trying to get the 6502X and the 65C02 in the green, and then I will re-evaluate how to proceed chopping the changes up into manageable chunks and start making PRs. A challenge there is that the changes that need to be done don't factor into incremental PRs as nicely as I would have liked them to.

**Comment by sidneycadot** [comment]
As to the 65C02 support: I traced the sim65 discrepancies with the test-set behavior to two things.

The first is a sim65 bug: it uses the same implementation of ADC and SBC for the 65C02 and the 6502. The 65C02 however changed the behavior of the processor status flags for ADC and SBC relative to the 6502, in case of decimal mode. This behavioral change is not implemented in sim65.

The second is a 65x02 testset bug: the test cases for the "BIT #imm" instruction that was added in the 65C02 appear to be incorrect in many cases. This issue was reported here:  SingleStepTests/65x02#10.

This second issue is a demonstration of the traceability/trust issue associated with the 65x02 test-set, that was touched upon before. I raised that issue in SingleStepTests/65x02#11.

**Comment by sidneycadot** [comment]
Ok, as to the differences between 6502X behavior in sim65, and the 65x02 testset:

There's a bug in the current sim65 where the handling of the "JMP (ind)" bug only happens for the 6502 CPU variant, not the 6502X variant. That's easily fixed.

For illegal opcode 0x8b, this depends on the "magic constant" being used that determines the instruction's behavior. The current implementation in sim65 uses 0xef, while the "No More Secrets" (https://csdb.dk/release/?id=238036) document doesn't list that particular value, but does list 0xee. If I change the sim65 emulator to use that, it behaves fully in accordance with the 65x02 testsuite, suggesting that magic constant value was used there, during generation of the testset (but: we don't know due to lack of traceability).

For the five illegal opcodes 0x93, 0x9B, 0x9C, 0x9E, 0x9F, I could not find small changes in the sim65 code that made them behave to pass the 65x02 test-suite. But I didn't put a lot of effort in it, given that it doesn't interest me a lot.

**Comment by sidneycadot** [comment]
Ok all the work so far has lead to a version of sim65's "6502.c" that is sitting on my disk with a bunch of fixes compared to the "6502.c" in the sim65 source tree. Upon diff'ing, this is a summary of the changes, from simple to complex:

(A) The fact that the JMP (ind) bug is not used for the 6502X CPU. A sim65 bug that's as clear as can be. I'll prepare a PR.
(B) The definition of the magic constant for the ANE illegal opcode (0x8b). Current implementation in sim65 is to use magic constant value 0xEF, we could change it to 0xEE, which would provide compliance with the 65x02 testsuite. I'll leave the decision about this to @mrdudz  who is more knowledgable about this.
(C) Handling of the "strange" (as in, not really existing) status bits 4 and 5 in the P "pseudo" register. I propose to open a separate issue for this to discuss, I have a proposal to change it. This would fix some minor bugs, where bad values get pushed to the stack.
(D) Several cases of clockcycle count errors in sim65, which I can now easily verify (at least for the regular, legal 6502 opcodes) using my own tool (sidneycadot/6502-test). The easy (and easily verified) fixes at least can be applied, I think.
(E) An esoteric corner case of JSR behavior if the instruction is located on the stack(!), and pushing the return address overwrites the target address. The 65x02 testsuite suggests a behavior there but I want to verify this on my real Atari.
(F) A number of simple changes that are needed to turn the simulator's registers as defined in CPURegs into uint8_t / uint16_t types, instead of the bare unsigned type. Fixing those types breaks some of the current opcode implementations, so those opcode implementations need to change to no longer depend on their type having extra bits. Mostly easy fixes that can by shown to be correct by reading the diffs.

I propose tackling this list slowly over the next few weeks, PR by PR. Each PR should have a clear-self-contained purpose, and should bring us closer to a sim65 where all known issues are resolved.

**Comment by mrdudz** [comment]
just a quick one (more later, i'm on the run...)

> The definition of the magic constant for the ANE illegal opcode (0x8b). Current implementation in sim65 is to use magic constant value 0xEF, we could change it to 0xEE, which would provide compliance with the 65x02 testsuite. I'll leave the decision about this to @mrdudz who is more knowledgable about this.

quoting myself from that doc: "For Emulation the best compromise between "proper emulation" and "making things work" seems to be to use a "magic contant" of $EF in regular cycles, and $EE in the RDY cycle."

> (E) An esoteric corner case of JSR behavior if the instruction is located on the stack(!), and pushing the return address overwrites the target address. The 65x02 testsuite suggests a behavior there but I want to verify this on my real Atari.

There is a test in the VICE testbench for this (verified on c64) -> https://sourceforge.net/p/vice-emu/code/HEAD/tree/testprogs/CPU/jsrselfmod/

**Comment by sidneycadot** [comment]
Hi @mrdudz 
> quoting myself from that doc[...]

Well, sim65 simulates at the instruction level rather than the cycle level, so here we are. Your call; I don't really care.

> There is a test in the VICE testbench for this

Thanks I'll look into that, saves me a bit of trouble.

**Comment by mrdudz** [comment]
> Well, sim65 simulates at the instruction level rather than the cycle level, so here we are. Your call; I don't really care.

It means that $ef is fine - there is no RDY line involved in sim65 at all :) (And really, it doesnt matter - $ee is just as good or bad. Any program relying on this is broken anyway)

**Comment by sidneycadot** [comment]
The 65C02 doesn't have illegal opcodes, anything that's not used is a NOP.

**Comment by sidneycadot** [comment]
As a small update: I received a Neo6502 board, which is built around a WDC 65C02 processor. This gives me both the means and the motivation to also fix the 65C02 issues still lurking in sim65.

The current 65C02 support od sim65 is a bit lacking; in the first place, sim65 doesn't support a number of 65C02-specific opcodes (declaring them as Illegal while they are, in fact, fully functional and documented). In the second place, it has errors in some implementations of opcodes. Most notably, the ADC and SBC instructions in the 65C02 behave differently in decimal mode compared to the 6502. Here's an example:

![image](https://github.com/user-attachments/assets/efbca9eb-43ff-439c-9bf7-c3fc18065f64)

The top row shows behavior of a hardware 6502 (as sampled on my Atari 800 XL); the bottom row shows behavior on a hardware 65C02 (my new Neo6502 board). They are clearly different; yet sim65 pretends that the 65C02 behaves identically to the 6502. I'm currently figuring out the precise behavior (reverse engineering from the sampled data, and roaming the internet for 65C02 simulators), but it will take some work. When I understand the differences, I intend to correct the 65C02 implementation.

**Comment by sidneycadot** [comment]
After studying the ADC and SBC functional behavior on the 6502 and the 65C02 (both of which I now have at my disposal for hardware-based tests), I implemented a C version of both opcodes, that is correct when checked against hardware. (See: https://github.com/sidneycadot/6502-test/tree/main/functional_test/adc_sbc/c_and_python_implementations).

Incorporating those new versions in my own version of sim65/src/6502.c, I now have a version where all currently implemented opcodes for the 65C02 are functionally correct:

![image](https://github.com/user-attachments/assets/9950b827-1340-41ca-a44c-41d09ab84c16)

The WDC6502 is not yet _perfect_ though. For that, we need two more things:

* Fix the instruction timings. The 65C02 deviates in cycle timing from the 6502 in a number of places; and it also includes ~ 100 new instructions that need to be tested. I want to verify this with a hardware implementation.
* Implement missing instructions. The table above shows opcodes marked as "illegal", but they are not; currently sim65 just doesn't implement them. The 65C02 has defined behavior for each and every opcode.

My personal version of sim65 core cpu emulation routine (which can be found here: https://github.com/sidneycadot/sim65-test-prototype)  is quite far ahead of CC65's version as it includes a lot of fixes that I plan to push back to the CC65 repo in a bunch of self-contained PRs).

I'm afraid I'm on a bit of a mission now to get sim65 as correct as it can be, with full traceability to hardware. Depending on the turnaround of code review and acceptance of the PRs that come out of this, this entire undertaking could take a few months.

**Comment by sidneycadot** [comment]
Okay more progress.

My clock cycle tester (https://github.com/sidneycadot/6502-test/tree/main/timing_test) runs on the Atari and the C64. It can now reproduce the behavior of a 6502: all 151 documented instructions, and most of the undocumented ones. Here's what that looks like on my Atari:

![image](https://github.com/user-attachments/assets/179e8bb3-8c58-4945-bcdf-fd16d864752f)

The 12 JAM instructions are obviously not testable, but there are more issues.

On my Atari, I have trouble testing 5 other undocumented opcodes: the 2 SHA instructions, as well as the SHX, the SHY, and the TAS pseudo-instructions. These five all cause similar mayhem while testing -- strange behavior, possibly some kind of memory corruption. I may hit upon previously unobserved corner cases of those instructions; I exercise a lot of addressing mode and index register combinations. Alternatively, my tester may still have a bug, but by now that's hard to believe, If I replace these instructions with well-behaved documented instructions that should have the same addressing mode behavior, the crashes don't happen.

I have ported my clock cycle tester to a 65C02, and it can check clock cycle counts for all 254 checkable opcodes on the 65C02:

![image](https://github.com/user-attachments/assets/547d1de1-0987-4269-a760-43f3f3080f70)

(The uncheckables on the 65C02 are the STP and WAI instructions, which have an indeterminate clock cycle duration.)

So. I've learned a lot, and I've also made good progress identifying where sim65 is not doing the right thing. I think it's all fixable, in essence, so that's promising.

... Which takes me to what I perceive to be a real issue: the lack of activity in the management of the CC65 repository. I posted two initial small PRs (#2555 and #2557), about two weeks ago, essentially as a warmup exercise, and they haven't been reviewed yet.

I think tackling all issues I now know of will take between 10 and 20 PRs, several of which will be considerably more complex than those two. It will be quite a lot of work to review that, but at the same time, it has the potential to seriously improve the quality of the sim65 simulator. Essentially, we are in a position to fix a dozen or so functional mismatches, as well as about a dozen or so timing (clock-cycle count) mismatches.

I'm ready to go, but this lack of activity is draining my motivation. Can we please discuss if and how to proceed on this? I've invested a lot of time by now (somewhere between 40 and 100 hourrs I think), it would be such a waste to throw all that away. But unless we can get to a more active process of dealing with the bug fixes that I have lined up, I will just abandon it all. No use flogging a dead horse, as the saying goes.

**Comment by mrdudz** [comment]
> ... Which takes me to what I perceive to be a real issue: the lack of activity in the management of the CC65 repository. I posted two initial small PRs (https://github.com/cc65/cc65/pull/2555 and https://github.com/cc65/cc65/pull/2557), about two weeks ago, essentially as a warmup exercise, and they haven't been reviewed yet.

I have merged some more PRs now. In any case - if you need a certain PR merged so you can proceed, please write a short reminder in that specific PR. If everything is silent, i might indeed not look for a week or two (in particular at busy periods of the year, like now :))

> I think tackling all issues I now know of will take between 10 and 20 PRs, several of which will be considerably more complex than those two. It will be quite a lot of work to review that, but at the same time, it has the potential to seriously improve the quality of the sim65 simulator. Essentially, we are in a position to fix a dozen or so functional mismatches, as well as about a dozen or so timing (clock-cycle count) mismatches.

it all looks very good so far, i agree

> I'm ready to go, but this lack of activity is draining my motivation. Can we please discuss if and how to proceed on this? I've invested a lot of time by now (somewhere between 40 and 100 hours I think), it would be such a waste to throw all that away. But unless we can get to a more active process of dealing with the bug fixes that I have lined up, I will just abandon it all. No use flogging a dead horse, as the saying goes.

I feel ya - if you want to move forward quickly, then the lack of active developers gets in the way. However, there is no solution to this other than "more active developers" :)

At this point i would usually offer to give you write access. You'll make PRs just like before, but you will be able to merge them yourself, perhaps after keeping them open for a day or two, or whatever makes sense. (And although you could, in theory, push directly to the main branch, please avoid doing this).

Obviously if we go this route, we must absolutely agree on that eg you respect the codestyle, or that you won't go wild and patch random parts of the toolchain just because you can ;) (i am also still offering this to @acqn - should you still be around)

**Comment by sidneycadot** [comment]
@mrdudz - I think the workflow where I produce self-contained PRs and self-approve them after a few days if nobody objects could work well. It allows me to pick a decent pace, while at the same time allowing others (such as yourself) to tap the brakes is you see things that are objectionable.

It is understood of course that I am not going to do silly stuff. My work will be limited to sim65 for starters, and i know that corner of the codebase pretty well be now. And while some choices concerning code style are different from what I would prefer, consistency is more important than anybody's personal taste.

**Comment by sidneycadot** [comment]
@mrdudz - I am looking into the behavior of the 5 undocumented opcodes that make my Atari misbehave. By now I'm very confident that the behavior I'm seeing is quite different than what is written up in the "no more secrets" document.

This is my test program. It produces a lot of BAD result an after a few dozens of tests it crashes the system: https://github.com/sidneycadot/6502-test/tree/main/functional_test/illegal_instructions. I suppose on a '64 it would just say "ok" for every test, and I will verify with a friend later this week.

I suppose that the "no more secrets" document was mostly written by C64 people using a 6510, whereas the Ataris have an Atari-customized 6502 ("SALLY"). So this could really be a behavioral divergence.

What is the proper place to discuss these matters?

**Comment by mrdudz** [comment]
We can discuss this via mail -> see the pdf ("wanted") :) the opcodes you are looking at are the "instable" ones - interesting that they behave so radically different on "SALLY" chips! we have tested these quite a lot on C64s, C128s and 1541s - but not so much on other platforms

**Comment by sidneycadot** [comment]
@mrdudz - I've posted a few PRs to make progress on the sim65 cleanup -- reviewing #2563 and #2564 would be really useful, as the weekend is coming up and getting these into the master branch will allow me to make nice progress.

I'd also appreciate if you can grant the previously discussed write access, and if we can agree on a policy w.r.t. the approx number of days you'd like to have PRs pending for comments, before it's reasonable to merge them to master.

In practice I think there should be a difference between stuff that's quite trivial (like #2564) and stuff that's quite involved (like #2563). In any case it would be pretty useful to get a heads-up in case you or other people consider a PR to be important enough to want to take time to do a bit of pondering -- a simple "please delay merging for now"message will suffice.

**Comment by mrdudz** [comment]
> reviewing https://github.com/cc65/cc65/pull/2563 and https://github.com/cc65/cc65/pull/2564 would be really useful

done

> I'd also appreciate if you can grant the previously discussed write access,

done :)

One thing i forgot: please label the issues/PRs before merging/closing (or even better - when creating :))

> and if we can agree on a policy w.r.t. the approx number of days you'd like to have PRs pending for comments, before it's reasonable to merge them to master.
> In practice I think there should be a difference between stuff that's quite trivial (like https://github.com/cc65/cc65/pull/2564) and stuff that's quite involved (like https://github.com/cc65/cc65/pull/2563). In any case it would be pretty useful to get a heads-up in case you or other people consider a PR to be important enough to want to take time to do a bit of pondering -- a simple "please delay merging for now"message will suffice.

As you say, its a bit hard to find a hard rule for this. mostly trivial stuff can be merged quickly, i guess. If in doubt, you might just request a comment :)

In any case - have fun :) Happy to see some more active development going on...

**Comment by mrdudz** [comment]
(That said, i probably won't have a lot of time for cc65 in the next week or two - ping me if there is something urgent)

**Comment by sidneycadot** [comment]
Much obliged!

With PRs #2567, #2568, and especially #2569 (which addresses the CPU register definitions with their proper bit-widths), we've reached a bit of a milestone: the simulator in 6502 and 65C02 mode is now functionally compliant with all tests of the 65x02 test suite, for the currently implemented instructions.

I will self-approve these changes in a few days unless anybody has questions/remarks.

After that, I will focus on getting the timing correct, not only wrt the 65x02 suite, but especially with my own tool that has been validated on hardware 6502 (Atari and C64) and the 65C02 by now.

Once that is done we're basically ready to close this big issue, as we'll have solved all easily solvable (and some not-so-easily-solvable) issues that came to light using the 65x02 tests. At that point, I will introduce a handful of new (much smaller) issues and feature requests concerning sim65 behavior, which I will then tackle myself (except, perhaps, the behavior of the handful of undocumented opcodes where sim65 and the 65x02 test suite disagree, at that point).

Those future issues will include stuff like:

- implementing the remaining 65C02 instructions (currently marked, incorrectly, as "illegal"). Preferably, I'd also want to include a reasonable implementation for the WAI and STP which are a bit weird.
- adding an option to sim65 to override the CPU variant encoded in the executable file.
- adding an option to sim65 to produce execution traces.
- adding a way to NOT have the instruction execution function at the heart of sim65 print stuff or terminate the program, which it currently can do under certain circumstances.
- figuring out an alternative for the paravirtualization stuff, which I feel is a bit messy.
- implementing regression tests for the instruction timings once they are working.
- implementing regression tests for ADC and BC behavior, which are by far the most complicated 65(C)02 instructions to get 100% right -- they are now, and let's keep it that way.
- figuring out what's going on with the few 6502 undocumented instructions where 65x02 and sim65 disagree (I can provide some help, but no promises there, for lack of deep interest. I like my undefined behavior, well, undefined.

**Comment by sidneycadot** [comment]
I just added two pull requests to improve sim65.

* #2572 adds implementations for 32 miscellaneous instructions. With these instructions added, sim65 now implements all 254 instructions that are testable against the 65x02 test suite (not testable: WAI and STP).
* #2573 fixes the implementation of 5 illegal instructions for the 6502X CPU mode. This concerns instructions that behave very weirdly when a page crossing is encountered; the previous implementation was completely wrong in that respect.

These two PRs once again bring sim65 closer to the target of being functionally complete and correct for all three CPU modes (6502 / 6502X / 65C02).

**Comment by sidneycadot** [comment]
Added PR #2575 that fixes all instruction timings in sim65, both for the 6502 and 65C02.

This basically completes this issue. One more issue-related PR is coming up, which is code style / code cleanup sweep over the sim65 code. A lot of changes have been made in the last few weeks, and going over the code to make sure the coding style is adhered to and so on is a good idea. I will do that after this PR is merged.

**Comment by sidneycadot** [comment]
Merged the last commit for this issue.

As of now, this is the result of the sim65 simulator when checked against the 65x02 test suite:

<img width="829" alt="image" src="https://github.com/user-attachments/assets/96844be9-da5f-42ad-a11c-b669e8dcec3f" />

The one deviation is the 0x5c opcode in the 65C02 tests. The 65x02 test suite claims that this instruction (purportedly, a NOP with absolute addressing mode) takes 4 cycles, but tests with a hardware WDC 65C02 have shown that this instruction takes 8 clock cycles. This deviation was reported to the 65x02 repository: https://github.com/SingleStepTests/65x02/issues/12

This concludes the work on fixing the sim65 simulator. I now believe it is as correct as an instruction-level simulator can be, both in terms documented and undocumented behavior, and in terms of instruction clock-cycle timing.

Closing the issue.

**Comment by mrdudz** [comment]
cool, thanks!
